plugins/catalog-backend/src/ingestion/processors/gitlab/groups.ts

```diff
@@ -0,0 +1,179 @@
+  _id: string,
+  pathDelimiter: string,
+  groupType: string = 'team',
+): Promise<GroupAdjacency> {
+  const groups = paginated<Group>(
+    options => client.pagedRequest('/groups', options),
+    { per_page: 100 },
+  );
+
+  const groupAdjacency = new Map<number, GroupNode>();
+  for await (const result of groups) {
+    const entity: GroupEntity = {
+      apiVersion: 'backstage.io/v1alpha1',
+      kind: 'Group',
+      metadata: {
+        name: result.full_path.replaceAll('/', pathDelimiter),
+      },
+      spec: {
+        type: groupType,
+        profile: {},
+        children: [],
+        members: [],
+      },
+    };
+    if (result.name) entity.spec!.profile!.displayName = result.name;
+    if (result.description) entity.metadata!.description = result.description;
+
+    if (!groupAdjacency.has(result.id)) {
+      groupAdjacency.set(result.id, {
+        children: [],
+        parent: result.parent_id,
+        entity,
+      });
+    }
+  }
+  await populateChildrenMembers(client, groupAdjacency);
+  mapChildrenToEntityRefs(groupAdjacency);
+  return groupAdjacency;
+}
+
+export async function populateChildrenMembers(
+  client: GitLabClient,
+  groupAdjacency: GroupAdjacency,
+) {
+  // map ids of children using parent for each group node
+  for (const [id, { parent, entity }] of groupAdjacency) {
+    // append as child of parent if present
+    if (parent !== null) {
+      const parentGroupNode = groupAdjacency.get(parent);
+      if (parentGroupNode) {
+        parentGroupNode.children.push(id);
+      }
+    }
+    // populate direct members
+    const users = await getGroupMembers(client, String(id), false);
+    for (const user of users) {
+      entity.spec!.members!.push(`user:${user.metadata.name}`);
+    }
+    // populate direct members of shared_with_groups
+    const sharedWithGroups = await getSharedWithGroupsIDs(client, String(id));
+    for (const sharedID of sharedWithGroups) {
+      // populate direct members
+      const sharedGroupUsers = await getGroupMembers(client, sharedID, false);
+      for (const user of sharedGroupUsers) {
+        entity.spec!.members!.push(`user:${user.metadata.name}`);
+      }
+    }
+  }
+}
+
+async function getSharedWithGroupsIDs(
+  client: GitLabClient,
+  id: string,
+): Promise<string[]> {
+  const response = await client.request(`/groups/${encodeURIComponent(id)}`);
+  const { shared_with_groups } = await response.json();
+  if (shared_with_groups) {
+    return shared_with_groups.map((group: SharedGroup) =>
+      String(group.group_id),
+    );
+  }
+  return [];
+}
+
+function mapChildrenToEntityRefs(groupAdjacency: GroupAdjacency) {
+  // map entity references of children from ids
+  for (const [_, { entity, children }] of groupAdjacency) {
+    for (const child of children) {
+      const childEntity = groupAdjacency.get(child)?.entity;
+      if (childEntity) {
+        entity.spec!.children!.push(
+          `group:default/${childEntity.metadata!.name}`,
```

Comment:
You probably want to use `stringifyEntityRef` from `@backstage/catalog-model` for all ref building, instead of doing it manually. You want to get the namespace and casing and stuff right